File: src/main/java/uk/co/thinkofdeath/patchtools/wrappers/ClassWrapper.java
/*
 * Copyright 2014 Matthew Collins
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uk.co.thinkofdeath.patchtools.wrappers;

import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.FieldNode;
import org.objectweb.asm.tree.MethodNode;

import java.util.ArrayList;
import java.util.List;

public class ClassWrapper {

    private final ClassNode node;
    private final ClassSet classSet;
    private final boolean hidden;
    private final List<MethodWrapper> methods = new ArrayList<>();
    private final List<FieldWrapper> fields = new ArrayList<>();

    public ClassWrapper(ClassSet classSet, ClassNode node) {
        this(classSet, node, false);
    }

    public ClassWrapper(ClassSet classSet, ClassNode node, boolean hidden) {
        this.classSet = classSet;
        this.node = node;
        this.hidden = hidden;

        node.methods.forEach(v -> methods.add(new MethodWrapper(this, v)));
        node.fields.forEach(v -> fields.add(new FieldWrapper(this, v)));
        if (hidden) {
            methods.forEach(v -> v.hidden = true);
            fields.forEach(v -> v.hidden = true);
        }
    }

    public ClassNode getNode() {
        return node;
    }

    public boolean isHidden() {
        return hidden;
    }

    public ClassSet getClassSet() {
        return classSet;
    }

    public List<MethodWrapper> getMethods() {
        return methods;
    }

    // Shouldn't ever been updated so we cache
    private MethodWrapper[] methodCache;

    public MethodWrapper[] getMethods(boolean stripHidden) {
        if (stripHidden) {
            if (methodCache == null) {
                methodCache = getMethods().stream()
                        .filter(m -> !m.isHidden())
                        .toArray(MethodWrapper[]::new);
            }
            return methodCache;
        }
        return getMethods().toArray(new MethodWrapper[getMethods().size()]);
    }

    public MethodNode getMethodNode(MethodWrapper wrapper) {
        MethodNode mn = node.methods.stream()
                .filter(m -> m.name.equals(wrapper.getName())
                        && m.desc.equals(wrapper.getDesc()))
                .findFirst().orElse(null);
        if (mn == null && node.superName != null) {
            mn = classSet.getClassWrapper(node.superName).getMethodNode(wrapper);
        }
        return mn;
    }

    public MethodWrapper getMethod(String name, String desc) {
        MethodWrapper wrap = methods.stream()
                .filter(m -> m.getName().equals(name)
                        && m.getDesc().equals(desc))
                .findFirst().orElse(null);
        if (wrap == null && node.superName != null) {
            wrap = classSet.getClassWrapper(node.superName).getMethod(name, desc);
        }
        return wrap;
    }

    public List<FieldWrapper> getFields() {
        return fields;
    }

    // Shouldn't ever been updated so we cache
    private FieldWrapper[] fieldCache;

    public FieldWrapper[] getFields(boolean stripHidden) {
        if (stripHidden) {
            if (fieldCache == null) {
                fieldCache = getFields().stream()
                        .filter(f -> !f.isHidden())
                        .toArray(FieldWrapper[]::new);
            }
            return fieldCache;
        }
        return getFields().toArray(new FieldWrapper[getFields().size()]);
    }

    public FieldWrapper getField(String name, String desc) {
        FieldWrapper wrap = fields.stream()
                .filter(f -> f.getName().equals(name)
                        && f.getDesc().equals(desc))
                .findFirst().orElse(null);
        if (wrap == null && node.superName != null) {
            wrap = classSet.getClassWrapper(node.superName).getField(name, desc);
        }
        return wrap;
    }

    public FieldNode getFieldNode(FieldWrapper fieldWrapper) {
        FieldNode fn = node.fields.stream()
                .filter(f -> f.name.equals(fieldWrapper.getName())
                        && f.desc.equals(fieldWrapper.getDesc()))
                .findFirst().orElse(null);
        if (fn == null && node.superName != null) {
            fn = classSet.getClassWrapper(node.superName).getFieldNode(fieldWrapper);
        }
        return fn;
    }

    @Override
    public String toString() {
        return "ClassWrapper{" + node.name + "}";
    }
}

File: src/main/kotlin/uk/co/thinkofdeath/patchtools/wrappers/ClassWrapper.kt
/*
 * Copyright 2014 Matthew Collins
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uk.co.thinkofdeath.patchtools.wrappers

import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.MethodNode
import org.objectweb.asm.tree.FieldNode

public class ClassWrapper(public val classSet: ClassSet, public val node: ClassNode, private val hidden: Boolean = false) {
    public val methods: MutableList<MethodWrapper> = arrayListOf()
    public val fields: MutableList<FieldWrapper> = arrayListOf();

    {
        node.methods.forEach {
            methods.add(MethodWrapper(this, it))
        }
        node.fields.forEach {
            fields.add(FieldWrapper(this, it))
        }
    }

    public fun isHidden(): Boolean {
        return hidden
    }

    // Shouldn't ever been updated so we cache
    private var methodCache: Array<MethodWrapper>? = null

    public fun getMethods(stripHidden: Boolean): Array<MethodWrapper> {
        if (stripHidden) {
            if (methodCache == null) {
                methodCache = methods
                    .filter { !it.isHidden() }
                    .copyToArray()
            }
            return methodCache!!
        }
        return methods.copyToArray()
    }

    public fun getMethodNode(wrapper: MethodWrapper): MethodNode? {
        var mn: MethodNode? = node.methods
            .filter { it.name == wrapper.name }
            .filter { it.desc == wrapper.desc }
            .first
        if (mn == null && node.superName != null) {
            val owner = classSet.getClassWrapper(node.superName)
            if (owner != null) {
                mn = owner.getMethodNode(wrapper)
            }
        }
        return mn
    }

    public fun getMethod(name: String, desc: String): MethodWrapper? {
        var wrap: MethodWrapper? = methods
            .filter { it.name == name }
            .filter { it.desc == desc }
            .first
        if (wrap == null && node.superName != null) {
            val owner = classSet.getClassWrapper(node.superName)
            if (owner != null) {
                wrap = owner.getMethod(name, desc)
            }
        }
        return wrap
    }

    // Shouldn't ever been updated so we cache
    private var fieldCache: Array<FieldWrapper>? = null

    public fun getFields(stripHidden: Boolean): Array<FieldWrapper> {
        if (stripHidden) {
            if (fieldCache == null) {
                fieldCache = fields
                    .filter { !it.hidden }
                    .copyToArray()
            }
            return fieldCache!!
        }
        return fields.copyToArray()
    }

    public fun getField(name: String, desc: String): FieldWrapper? {
        var wrap: FieldWrapper? = fields
            .filter { it.name == name }
            .filter { it.desc == desc }
            .first
        if (wrap == null && node.superName != null) {
            val owner = classSet.getClassWrapper(node.superName)
            if (owner != null) {
                wrap = owner.getField(name, desc)
            }
        }
        return wrap
    }

    public fun getFieldNode(fieldWrapper: FieldWrapper): FieldNode? {
        var fn: FieldNode? = node.fields
            .filter { it.name == fieldWrapper.name }
            .filter { it.desc == fieldWrapper.desc }
            .first
        if (fn == null && node.superName != null) {
            val owner = classSet.getClassWrapper(node.superName)
            if (owner != null) {
                fn = owner.getFieldNode(fieldWrapper)
            }
        }
        return fn
    }

    override fun toString(): String {
        return "ClassWrapper{" + node.name + "}"
    }
}